src/scanpy/preprocessing/_simple.py

```diff
@@ -722,13 +736,13 @@ def regress_out(
                 "we regress on the mean for each category."
             )
         logg.debug("... regressing on per-gene means within categories")
-        regressors = np.zeros(X.shape, dtype="float32")
+        # Create numpy array's from categorical variable
+        number_categories = np.int64(len(adata.obs[keys[0]].cat.categories))
+        filters = adata.obs[keys[0]].cat.codes.to_numpy()
+        number_categories = number_categories.astype(filters.dtype)
```

Comment:
Either this or add a comment (to the code) explaining why it needs to be the other way

```suggestion
        number_categories = len(adata.obs[keys[0]].cat.categories)
        filters = adata.obs[keys[0]].cat.codes.to_numpy()
```

Reply:
I added a comment. Other wise you have a dtype missmatch and crash of the kernel

Reply:
> Other wise you have a dtype missmatch and crash of the kernel

I would say that this is the important part for the comment!

Reply:
100%!

- refactor your code until the “what” is obvious.
- if the “why” isn’t obvious from understanding the “what”, add the missing parts as a comment

I see that you’re

1. convert the cat codes into a numpy array
2. creating a numpy scalar with the same dtype as `filters`, holding the number of categories

So  you don’t need to comment that you do any of that.

I asked because I’m confused why a Python integer is converted to a numpy scalar: Usually APIs accept either and do the converting themselves. So I’d like to see a comment removing that confusion by explaining *why* you convert to a numpy scalar. (a crash is a great reason)

----

but I also see that `_create_regressor_categorical` has `number_categories: int` and then does `range(number_categories)`, so I’m still very confused why numba crashes unless the dtypes match.

I can’t reproduce the crash. leaving the thing as a Python int just works for me.

Reply:
Also the way to do this in one step is

```suggestion
        filters = adata.obs[keys[0]].cat.codes.to_numpy()
        number_categories = filters.dtype.type(len(adata.obs[keys[0]].cat.categories))
```

Reply:
That worked.